harness/determined/_info.py

```diff
@@ -0,0 +1,371 @@
+import json
+import os
+from typing import Any, Dict, List, Optional
+
+from determined import gpu
+
+
+def getenv_int(key: str) -> Optional[int]:
+    val = os.environ.get(key)
+    return val if val is None else int(val)
+
+
+class RendezvousInfo:
+    """
+    RendezvousInfo is machine identity information that is:
+     - configured in the container by the rendezvous layer (when on a Determined cluster)
+     - independent of the launch layer
+     - consumed by the launch layer
+    """
+
+    def __init__(self, addrs: List[str], rank: int):
+        self._addrs = addrs
+        self._rank = rank
+
+    @property
+    def container_ips(self) -> List[str]:
+        return self._addrs
+
+    @property
+    def container_rank(self) -> int:
+        return self._rank
+
+    def _to_file(self, path: str = "/run/determined/info/rendezvous.json") -> None:
+        """
+        to_file writes the RendezvousInfo to a well-known location in a Determined container.  This
+        is called automatically early in the lifetime of a container, so user code can always expect
+        this file to be written before user code runs.
+        """
+        with open(path, "w") as f:
+            json.dump({k.lstrip("_"): v for k, v in vars(self).items()}, f)
+
+    @classmethod
+    def _from_file(
+        cls,
+        path: str = "/run/determined/info/rendezvous.json",
+    ) -> Optional["RendezvousInfo"]:
+        if not os.path.exists(path):
+            return None
+        with open(path) as f:
+            return cls(**json.load(f))
+
+
+class TrialInfo:
+    def __init__(
+        self,
+        trial_id: int,
+        experiment_id: int,
+        trial_seed: int,
+        hparams: Dict[str, Any],
+        config: Dict[str, Any],
+    ):
+        self._trial_id = trial_id
+        self._experiment_id = experiment_id
+        self._trial_seed = trial_seed
+        self._hparams = hparams
+
+        # _config has no getter because it's not a stable API; as the experiment config version
+        # increases, the shape of the experiment config dict might change as the master is upgraded.
+        # As a result, we should strongly discourage users from reading the experiment config dict
+        # in their training code.  They should focus on the data field and the hparams field, which
+        # have always been and will always be user-defined.
+        self._config = config
+
+    @property
+    def trial_id(self) -> int:
+        return self._trial_id
```

Comment:
I have increasingly seen external libraries that use the `@property` decorator in all cases, so I'm not sure that it's exactly _un_-pythonic.

That said, I'm not sold on it myself.  The cons I see are:
- I don't like the extra verbosity.
- It takes something simple and makes it not simple

But the pros I see are:
- I do like the read-only nature of properties; it shows our intentions for the API very clearly, if any users ever casually try to set it or if they read the code.  Users who read the source code to find `._trial_id` and set that will eventually get what they had coming when their code breaks... there's no preventing that really.
- I like that us having a function getter is more flexible in the future, if we e.g. want to rename a property, and leave the old symbol present but with a deprecation warning.  And in that potential future case, it's nice that the property was read-only since the beginning.

I think the pros outweigh the cons.  I especially like the read-only-ness because in compiled languages I like `const` whenever possible.  But I don't think I care all that much, so if you feel strongly I can be ok with normal attributes rather than `@property` attributes.  Just let me know.

Reply:
>I like that us having a function getter is more flexible in the future, if we e.g. want to rename a property, and leave the old symbol present but with a deprecation warning. And in that potential future case, it's nice that the property was read-only since the beginning.

if you have a plain simple field `self.trial_id = trial`, and you want to deprecate it, you can add the `@property` getter for it specifically with the deprecation warning at that time. making them `@properties` ahead of time doesn't add to futureproofing.

>Users who read the source code to find ._trial_id and set that will eventually get what they had coming when their code breaks... there's no preventing that really.

As a source code reader for 3p library code, I'd prefer cleaner and concise code, and I won't make assumptions it's safe to write into it unless the documentation says so.

Read-only-ness doesn't cut it for me. I'd vote against properties.